File: govent/day22/day22.go
package main

import (
	"fmt"
	"strings"
	"strconv"
	"os"
)

type Cube struct {
	x int
	y int
	z int
}

type Reactor struct {
	on_cubes map[Cube]struct{}
}

func (reactor *Reactor) turn_on_cuboid(xmin, xmax, ymin, ymax, zmin, zmax int) {
	for x := xmin; x <= xmax; x++ {
		for y := ymin; y <= ymax; y++ {
			for z := zmin; z <= zmax; z++ {
				reactor.on_cubes[Cube{x, y, z}] = struct{}{}
			}
		}
	}
}

func (reactor *Reactor) turn_off_cuboid(xmin, xmax, ymin, ymax, zmin, zmax int) {
	for x := xmin; x <= xmax; x++ {
		for y := ymin; y <= ymax; y++ {
			for z := zmin; z <= zmax; z++ {
				delete(reactor.on_cubes, Cube{x, y, z})
			}
		}
	}
}

type Instruction struct {
	on bool
	cuboid []int
}

func (reactor *Reactor) central_reboot(instructions []Instruction) {
	for _, instruction := range instructions {
		xmin := instruction.cuboid[0]
		xmax := instruction.cuboid[1]
		ymin := instruction.cuboid[2]
		ymax := instruction.cuboid[3]
		zmin := instruction.cuboid[4]
		zmax := instruction.cuboid[5]
		if xmin < -50 || xmax > 50 || ymin < -50 || ymax > 50 || zmin < -50 || zmax > 50{
			continue
		}
		if instruction.on {
			reactor.turn_on_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
		} else {
			reactor.turn_off_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
		}
	}
}

func (reactor *Reactor) full_reboot(instructions []Instruction) {
	for _, instruction := range instructions {
		fmt.Println(instruction)

		xmin := instruction.cuboid[0]
		xmax := instruction.cuboid[1]
		ymin := instruction.cuboid[2]
		ymax := instruction.cuboid[3]
		zmin := instruction.cuboid[4]
		zmax := instruction.cuboid[5]
		if instruction.on {
			reactor.turn_on_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
		} else {
			reactor.turn_off_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
		}
	}
}

func parse_input(source string) []Instruction {
	instructions := []Instruction{}
	for _, line := range strings.Split(source, "\n") {
		coords := []int{}
		l := strings.Split(line, " ")
		state := l[0] == "on"
		for _, coord := range strings.Split(l[1], ",") {
			coord_mini := strings.Split(coord[2:], "..")
			mi, _ := strconv.Atoi(coord_mini[0])
			ma, _ := strconv.Atoi(coord_mini[1])
			coords = append(coords, mi)
			coords = append(coords, ma)
		}
		instructions = append(instructions, Instruction{state, coords})
	}
	return instructions
}

func main() {
	dat, _ := os.ReadFile("day22/day22")
	s := string(dat)
	// s := "on x=10..12,y=10..12,z=10..12"

	instructions := parse_input(s)

	reactor := Reactor{make(map[Cube]struct{})}
	reactor.full_reboot(instructions)
	fmt.Println(len(reactor.on_cubes))
}

File: python/aoc2021/day22.py
import os
import itertools

class Cube():
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def __repr__(self):
        return f'Cube=({self.x} {self.y} {self.z})'

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y and self.z == other.z

    def __hash__(self):
        return hash((self.x, self.y, self.z))


class Reactor():
    def __init__(self):
        self.on_cubes = set()

    def make_cuboid(self, xmin, xmax, ymin, ymax, zmin, zmax):
        return {Cube(x,y,z) for x,y,z in itertools.product(range(zmin, zmax+1), range(ymin, ymax+1), range(xmin, xmax+1))}

    def turn_on_cuboid(self, xmin, xmax, ymin, ymax, zmin, zmax):
        print('making')
        cuboid = self.make_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
        print('doing')
        self.on_cubes.update(cuboid)

    def turn_off_cuboid(self, xmin, xmax, ymin, ymax, zmin, zmax):
        cuboid = self.make_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
        self.on_cubes.difference_update(cuboid)

    def central_reboot(self, instructions):
        for i in instructions:
            xmin, xmax, ymin, ymax, zmin, zmax = i[1:]
            if xmin < -50 or xmax > 50 or ymin < -50 or ymax > 50: continue
            if i[0] == 'on':
                self.turn_on_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
            elif i[0] == 'off':
                self.turn_off_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)

    def full_reboot(self, instructions):
        for i in instructions:
            print(i)
            xmin, xmax, ymin, ymax, zmin, zmax = i[1:]
            # if xmin < -50 or xmax > 50 or ymin < -50 or ymax > 50: continue
            if i[0] == 'on':
                self.turn_on_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)
            elif i[0] == 'off':
                self.turn_off_cuboid(xmin, xmax, ymin, ymax, zmin, zmax)


class InputParser():
    def __init__(self, s):
        self.lines = s.splitlines()
        self.instructions = []

    def parse(self):
        instructions = []
        for line in self.lines:
            instruction = []
            state, coords = line.split(' ')
            instruction.append(state)
            for coord in coords.split(','):
                coord = coord[2:]
                mi, ma = coord.split('..')
                instruction.append(int(mi))
                instruction.append(int(ma))
            instructions.append(instruction)
        return instructions


if __name__ == '__main__':
    with open(os.path.dirname(__file__) + '/res/day22') as f:
        parser = InputParser(f.read())
        instructions = parser.parse()
        reactor = Reactor()
        reactor.full_reboot(instructions)
        print(len(reactor.on_cubes))
